rfcs/0004-fulltext-search.md

```diff
@@ -0,0 +1,275 @@
+          { name: "title", weight: 5 },
+        ]
+      },
+      {
+        entity: "Musician",
+        fields: [
+          { name: "name", weight: 10 },
+          { name: "bio", weight: 5 },
+        ]
+      }
+    ]
+  )
+```
+
+The schema generated from the above definition: 
+```graphql
+union FulltextMediaResultItem = ...
+union FulltextSearchResultItem = Band | Album | Musician
+type Query {
+  media...
+  search(text: String!, first: Int, skip: Int, block: Int): [FulltextSearchResultItem!]!
+}
+```
+
+### GraphQL Query interface
+
+End users of the subgraph will have access to the queries generated for the
+fulltext search indexes alongside the other queries generated for each 
+entity in the subgraph.  In the case of a fulltext search index that spans
+multiple entities, [inline fragments](https://graphql.org/learn/queries/#inline-fragments) 
+may be used in the query to deal with the polymorphic result items. In the 
+front-end the `__typename` in each result item can be used to render the 
+components accordingly.  
+
+In the "text" parameter supplied to the query there will be one special 
+operator supported, the proximity operator. Several operators are available: 
+and, or, and proximity (`&`, `|`, `<->`.) The proximity operator allows 
+one to specify max distance between search terms: 3 words apart → `<3>`. 
+
+Example query using inline fragments and the proximity operator: 
+```graphql
+query {
+  search(text: "Bob<3>run") {
+    __typename
+    ... on Band { name label { id } }
+    ... on Album { title numberOfTracks }
+    ... on Musician { name bio }
+  }
+}
+```
+
+### Tools and Design
+
+Fulltext search query system implementations often involve specific systems 
+for storing and querying the text documents; however, in an effort to reduce 
+system complexity and feature implementation time I propose starting with the 
+fulltext search features built in to PostgreSQL.
+
+A FullText search field will get its own column in the Entity table just 
+like any other field; however, the data stored will be the result of the 
+lexical, morphological analysis of text documents performed on the fields 
+included in the index. The fulltext search field will be created using the 
+Postgres ts_vector function and will be indexed using a GIN index. The subgraph 
+developer will define a ranking algorithm to be used to sort query results,
+so the end-user facing API remains easy to use without any requirement to 
+understand the ranking algorithms.   
+
+Limitations of Postgres fulltext search indexes and search algorithms: 
+  - only certain languages available (expandable with plugins like [PGroonga](https://pgroonga.github.io/)), 
+    - **languages supported out of the box**: 
+        Danish, Dutch, English, Finnish, French, German, Hungarian, Italian, 
+        Norwegian, Portuguese, Romanian, Russian, Spanish, Swedish, and Turkish.
+  - select algorithms are available, and 
+  - limited to PostgreSQL storage. 
```

Comment:
Do we care about Postgres here? Isn't it an implementation detail? Aren't we starting only with English initially?

Reply:
I agree the Postgres implementation details should be in the plan not here.  

I remember starting with English only being mentioned on the PR discussion call, but I thought it was just an option in case we need to simplify. Is there any reason not to include the option for other languages if it doesn't require more implementation work?